### PR TITLE
Update hubert implementation.

### PR DESCRIPTION
**Pull request by simpleoier**
Here is a tentative PR for HuBERT implementation.

### Label quality
Label quality between MFCC Kmeans centroid against Librispeech Phoneme Alignments.
**Phone-normalized mutual information (PNMI)**, which is MI/H(ref) in the following table, for MFCC achieves a similar level as in Table 4 the [original HuBERT paper](https://arxiv.org/pdf/2106.07447.pdf)
```
Stage 6: Measure qualities of pseudo labels (nclusters=100)
  ref pur    hyp pur    H(ref)    H(hyp)      MI    MI/H(ref)    ref segL    hyp segL  p_xy shape      frm tot    frm diff    utt tot    utt miss
---------  ---------  --------  --------  ------  -----------  ----------  ----------  ------------  ---------  ----------  ---------  ----------
   0.2907     0.0920    3.4040    4.5434  0.8774       0.2578      9.5901      2.1890  (42, 100)       3721308         339       5567          16
```

### HuBERT Training (8 x A100 40GB)
* Final ACC_m: 46.0% / 46.5% @ 250epoch
* Pretrained model: [HuggingFace](https://huggingface.co/espnet/simpleoier_librispeech_hubert_iter0_train_ssl_torchaudiohubert_base_960h_pretrain_it0_raw)

|                           |      acc_m      |      acc_u      |         loss      |
| :-------------: | :------------: | :----------: | :----------: |
|Training curves | image | image |  image |

### ASR Finetuning

|         init_ckpt        |            Data          | WER (dev-clean / dev-other / test-clean / test-other | Final CTC loss | CTC Loss curve |
| :----------------: | :---------------: | :---------------------------------------------------: | :------------: | :---------------: |
| valid.loss.ave.pth | Librilight_limited | 15.5 / 23.9 / 15.9 / 24.1 | 63.272 (train) / 23.009 (valid) | image |
| valid.loss.best.pth | Librilight_limited | 15.3 / 24.2 / 15.7 / 24.8 | 65.477 (train) / 22.684 (valid) | image |
| latest.pth (250ep) | Librilight_limited | 15.5 / 24.3 / 15.9 / 24.8 | 64.145 (train) / 23.341 (valid) | image |

### Label quality
Label quality between **Iter0 Layer 6** Kmeans centroid against Librispeech Phoneme Alignments.
**Phone-normalized mutual information (PNMI)**, which is MI/H(ref) in the following table.
```
Stage 6: Measure qualities of pseudo labels (nclusters=500)
  ref pur    hyp pur    H(ref)    H(hyp)      MI    MI/H(ref)    ref segL    hyp segL  p_xy shape      frm tot    frm diff    utt tot    utt miss
---------  ---------  --------  --------  ------  -----------  ----------  ----------  ------------  ---------  ----------  ---------  ----------
   0.5988     0.0923    3.4040    6.0827  2.0930       0.6148      9.5901      3.1200  (42, 500)       3721308        3096       5567          16
```

### HuBERT Training (8 x A100 40GB)
* Final ACC_m: 51.8% / 47.0% (250epoch)
* Pretrained model: [Huggingface](https://huggingface.co/espnet/simpleoier_librispeech_hubert_iter1_train_ssl_torchaudiohubert_base_960h_pretrain_it1_raw)

|                           |      acc_m      |      acc_u      |         loss      |
| :-------------: | :------------: | :----------: | :----------: |
|Training curves | image | image | image |

### ASR Finetuning
* Pretrained model: [Huggingface](https://huggingface.co/espnet/simpleoier_librilight_limited_asr_train_asr_hubert_base_10h_finetuning_raw_en_char)

|         init_ckpt        |            Data          | WER (dev-clean / dev-other / test-clean / test-other | Final CTC loss | CTC Loss curve |
| :----------------: | :---------------: | :---------------------------------------------------: | :------------: | :---------------: |
| valid.loss.ave.pth | Librilight_limited | 10.4 / 17.4 / 10.5 / 17.6 | 50.133 / 15.069 | image |

## Some bugs we noticed in torchaudio. Please check these if some error encountered. 
* `no attribute kernel_size`  https://github.com/pytorch/audio/issues/2845
* `amp dtype mismatch`: https://github.com/pytorch/audio/issues/2847

## Todos
- [x] test scripts
- [x] results and pretrained models

### CODE REVIEW

**Comment by pyf98** [comment]
Line 89 in `data.sh`:

The zip file is downloaded to a specific folder.

```
unzip "${alignment_phoneme_dir}/librispeech_alignments.zip" -d "${alignment_phoneme_dir}"
```

**Comment by pyf98** [comment]
For the `collect stats`, I tried to set `--batch_size 1` here:
https://github.com/simpleoier/espnet/blob/hubert/egs2/TEMPLATE/ssl1/hubert.sh#L481
Otherwise, the audio lengths will be cropped to the minimum length in a mini-batch during `collect stats`, leading to wrong stats.

**Comment by sw005320** [comment]
LGTM!
Thanks for your great efforts.

Some comments (Please reflect them in the follow-up PR)
- Add (update) a description of this function in the main `README.md`
- Add (update) acknowledgments (Thanks @nateanl)

**Comment by wanchichen** [comment]
We may also want to put a check in `path.sh` or comment in the `README.MD` about the torchaudio version. torchaudio.models.hubert_pretrain_model is only available with a version >= 0.11.0. The AMP issues raised by @pyf98 are also not fixed until version 0.13.1.

**Comment by nateanl** [comment]
This is true. I suggest checking if torchaudio is 0.13.1 or nightly version, not only because of the AMP issue, but also there is new model initialization which is important to HuBERT pre-training.
